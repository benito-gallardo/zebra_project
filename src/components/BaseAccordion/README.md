# BaseAccordion Component

A global accordion component with default styling.

## How to use

```
import React from 'react';

import BaseAccordion from './component/BaseAccordion';

...
<BaseAccordion top="Lorem Ipsum" bottom="Lorem ipsum dolor adipiscing" />
```

## Default Props

| Prop Name     | Options           | Default Value  | Description                                                                                                                                                     |
| ------------- | ----------------- | -------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| top           | string, component | `Title`        | Sets the content in the `top` section of the accordion.                                                                                                         |
| bottom        | string, component | `Some content` | Sets the content in the `bottom` section of the accordion                                                                                                       |
| arrowPosition | string            | `right`        | `right` = Sets the arrow position to the right side of the `top` section <br/> `bottom` = Sets the arrow position to under the `bottom` section and centers it. |
| className     | string            | `accordion`    | Adds additional classes to the element. <br /> The default class is not overwriten.                                                                             |
