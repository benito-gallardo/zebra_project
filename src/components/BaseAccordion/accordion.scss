@import '../../assets/sass/vars/index';
.accordion {
  border: 1px solid $grey-light;
  padding: 1rem 3rem;
  margin: 1rem 0;
  cursor: pointer;
  .accordion-top {
    color: black;
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
  .accordion-icon {
    display: inline-block;
    cursor: pointer;
    &.bottom {
      display: block;
      margin: 0 auto;
    }
    polygon {
      fill: $grey;
    }
  }
  .accordion-bottom {
    max-height: 0;
    overflow: hidden;
    transition: max-height 0.5s ease;
    .accordion-content {
      padding-top: 1rem;
    }
  }
  &.active {
    .accordion-icon {
      transform: rotate(180deg);
    }
  }
}
