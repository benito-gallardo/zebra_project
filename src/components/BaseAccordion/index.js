import React, { useState, useRef } from "react";

import { ReactComponent as IconChevronDown } from '../../assets/images/IconChevronDown.svg';

import './accordion.scss';

const BaseAccordion = ({top='Title', bottom='Some content', arrowPosition= 'right', className=''}) => {
  const [activeState, setActiveState] = useState(false);
  const [heightState, setHeightState] = useState("0px");

  const content = useRef(null);

  function toggleAccordion() {
    setActiveState(prevActiveState => !prevActiveState);
    setHeightState(activeState ? "0px" : `${content.current.scrollHeight}px`);
  }

  return (
    <section className={`accordion ${activeState ? 'active' : ''} ${className}`} onClick={toggleAccordion} data-testid="base-accordion">
      
      <div className="accordion-top" data-testid="base-accordion-top">
        {top}
        {arrowPosition === 'right' && <IconChevronDown  className="accordion-icon right" data-testid="base-accordion-icon" />}
      </div>

      <div className={`accordion-bottom`} ref={content} style={{ maxHeight: `${heightState}`, visibility: `${activeState ?'visible' : 'hidden'}` }} data-testid="base-accordion-bottom" >
       <div className="accordion-content"> {bottom}</div>
      </div>

      {arrowPosition === 'bottom' &&<IconChevronDown className="accordion-icon bottom" data-testid="base-accordion-icon"/>}
    </section>
  )
};

export default BaseAccordion;