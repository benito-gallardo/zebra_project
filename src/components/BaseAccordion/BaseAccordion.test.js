import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import React from 'react';
import ReactDOM from 'react-dom';

import BaseAccordion from './index';

let Accordion;
afterEach(cleanup);

const props = {
  top: 'Top Content',
  bottom: 'Bottom Content',
  arrowPosition: 'bottom',
  className: 'test-class'
};

describe('The BaseAccordion Component', () => { 
  
  it('renders without crashing', () => { 
    const div = document.createElement('div');
    ReactDOM.render(<BaseAccordion />, div);
  })

  it('matches snapshot with props', () => { 
    const { asFragment } = render(<BaseAccordion {...props} />);
    expect(asFragment()).toMatchSnapshot();
  })

  it('matches snapshot without props', () => { 
    const { asFragment } = render(<BaseAccordion />);
    expect(asFragment()).toMatchSnapshot();
  })

  describe('when props are provided', () => { 

    beforeEach(() => { 
      const { getByTestId } = render(<BaseAccordion {...props} />);
      Accordion = getByTestId('base-accordion');
    })

    it('custom class is added', () => { 
      expect(Accordion).toHaveClass('accordion','test-class');
    })

    it('top and bottom content should be from props ', () => { 
      expect(Accordion).toHaveTextContent('Top Content');
      expect(Accordion).toHaveTextContent('Bottom Content');
    })

    it('has arrow icon at bottom of card', () => { 
      const Icon = screen.getByTestId('base-accordion-icon');
      expect(Icon).not.toHaveClass('right');
      expect(Icon).toHaveClass('bottom');
    })
  })

  describe('when props are not provided', () => { 

    beforeEach(() => { 
      const { getByTestId } = render(<BaseAccordion />);
      Accordion = getByTestId('base-accordion');
    })

    it('has default class', () => { 
      expect(Accordion).toHaveClass('accordion');
    })

    it('top and bottom content should be from props ', () => { 
      expect(Accordion).toHaveTextContent('Title');
      expect(Accordion).toHaveTextContent('Some content');
    })

    it('has arrow icon on the right side of card', () => { 
      const Icon = screen.getByTestId('base-accordion-icon');
      expect(Icon).not.toHaveClass('bottom');
      expect(Icon).toHaveClass('right');
    })
  })

  describe('when accordion is not open', () => {

    beforeEach(() => { 
      const { getByTestId } = render(<BaseAccordion />);
      Accordion = getByTestId('base-accordion');
    })

    it('does not have an active class', () => { 
      expect(Accordion).not.toHaveClass('active');
    })

    it('accordion content area is not visable', () => { 
      const AccordionBottom = screen.getByTestId('base-accordion-bottom');
      expect(AccordionBottom).not.toBeVisible();
    })
  })
    
  describe('when accordion is clicked', () => { 

    beforeEach(() => { 
      const { getByTestId } = render(<BaseAccordion />);
      Accordion = getByTestId('base-accordion');
      fireEvent.click(Accordion);
    })

    it('accordion classes update to active', () => { 
      expect(Accordion).toHaveClass('accordion', 'active');
    })

    it('accordion content area is visable', () => { 
      const AccordionBottom = screen.getByTestId('base-accordion-bottom');
      expect(AccordionBottom).toBeVisible();
    })
  })
})
