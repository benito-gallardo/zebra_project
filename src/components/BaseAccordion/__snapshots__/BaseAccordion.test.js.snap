// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The BaseAccordion Component matches snapshot with props 1`] = `
<DocumentFragment>
  <section
    class="accordion  test-class"
    data-testid="base-accordion"
  >
    <div
      class="accordion-top"
      data-testid="base-accordion-top"
    >
      Top Content
    </div>
    <div
      class="accordion-bottom"
      data-testid="base-accordion-bottom"
      style="max-height: 0px; visibility: hidden;"
    >
      <div
        class="accordion-content"
      >
         Bottom Content
      </div>
    </div>
    <svg
      class="accordion-icon bottom"
      data-testid="base-accordion-icon"
    >
      IconChevronDown.svg
    </svg>
  </section>
</DocumentFragment>
`;

exports[`The BaseAccordion Component matches snapshot without props 1`] = `
<DocumentFragment>
  <section
    class="accordion  "
    data-testid="base-accordion"
  >
    <div
      class="accordion-top"
      data-testid="base-accordion-top"
    >
      Title
      <svg
        class="accordion-icon right"
        data-testid="base-accordion-icon"
      >
        IconChevronDown.svg
      </svg>
    </div>
    <div
      class="accordion-bottom"
      data-testid="base-accordion-bottom"
      style="max-height: 0px; visibility: hidden;"
    >
      <div
        class="accordion-content"
      >
         Some content
      </div>
    </div>
  </section>
</DocumentFragment>
`;
